### PR TITLE
Add openHns method

### PR DESCRIPTION
**Pull request by mrcnski**

### CODE REVIEW

File: src/download.js [hunk 1]
```diff
@@ -1,35 +1,79 @@
...
   const url = this.getDownloadUrl(skylink, opts);
 
   window.open(url, "_blank");
 };
 
+SkynetClient.prototype.downloadHns = async function (hns, customOptions = {}) {
+  const opts = { ...defaultDownloadHnsOptions, ...customOptions };
+};
+
+SkynetClient.prototype.downloadHnsres = async function (hnsres, customOptions = {}) {
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
SkynetClient.prototype.resolveHns = async function (domainName, customOptions = {}) {
```

File: src/download.js [hunk 2]
```diff
@@ -1,35 +1,79 @@
...
+ * @param {string} [customOptions.endpointPath="/"] - The relative URL path of the portal endpoint to contact.
+ */
 SkynetClient.prototype.download = function (skylink, customOptions = {}) {
   const opts = { ...defaultDownloadOptions, ...customOptions, download: true };
   const url = this.getDownloadUrl(skylink, opts);
 
   window.open(url, "_blank");
 };
 
+SkynetClient.prototype.downloadHns = async function (hns, customOptions = {}) {
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
SkynetClient.prototype.downloadHns = async function (domainName, customOptions = {}) {
```

File: src/download.js [hunk 3]
```diff
@@ -1,35 +1,79 @@
...
+};
+
 SkynetClient.prototype.getDownloadUrl = function (skylink, customOptions = {}) {
   const opts = { ...defaultDownloadOptions, ...customOptions };
   const query = opts.download ? { attachment: true } : {};
 
-  return makeUrlWithSkylink(this.portalUrl, opts.endpointPath, skylink, query);
+  let url = makeUrl(this.portalUrl, opts.endpointPath, parseSkylink(skylink));
+  url = addUrlQuery(url, query);
+  return url;
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
  const url = makeUrl(this.portalUrl, opts.endpointPath, parseSkylink(skylink));
  
  return addUrlQuery(url, query);
```

File: src/download.test.js [hunk 4]
```diff
@@ -11,6 +16,7 @@ const validSkylinkVariations = [
   `${portalUrl}/${skylink}/foo/bar`,
   `${portalUrl}/${skylink}?foo=bar`,
 ];
+const validHnsLinkVariations = [hnsLink, `hns:${hnsLink}`, `hns://${hnsLink}`];
```
Comment (interstella-5555) [inline src/download.test.js:22]: I don't think those `hns:` and `hns://` are valid...

File: src/upload.js [hunk 5]
```diff
@@ -50,7 +50,8 @@ SkynetClient.prototype.uploadDirectory = async function (directory, filename, cu
     formData.append(opts.portalDirectoryFileFieldname, ensureFileObjectConsistency(file), path);
   });
 
-  const url = makeUrl(this.portalUrl, opts.endpointPath, { filename });
+  let url = makeUrl(this.portalUrl, opts.endpointPath);
+  url = addUrlQuery(url, { filename });
```
Comment (interstella-5555) [inline src/upload.js]: could you add query argument to makeUrl so it's all in one function?

File: src/utils.js [hunk 6]
```diff
@@ -1,9 +1,20 @@
 // import axios from "axios";
 import path from "path-browserify";
 import parse from "url-parse";
+import urljoin from "url-join";
 
 export const defaultSkynetPortalUrl = "https://siasky.net";
 
+export const uriHandshakePrefix = "hns:";
+export const uriHandshakeResolverPrefix = "hnsres:";
```
Comment (interstella-5555) [inline src/utils.js:9]: Are those valid? I haven't seen or used those myself

File: src/utils.js [hunk 7]
```diff
@@ -56,27 +67,23 @@ export function getRootDirectory(file) {
...
-  return makeUrl(portalUrl, path.posix.join(endpointPath, parsedSkylink), query);
+/**
+ * Properly joins paths together to create a URL. Takes a variable number of
+ * arguments.
+ */
+export function makeUrl() {
+  let args = Array.from(arguments);
+  return args.reduce(function (acc, cur) {
+    return urljoin(acc, cur);
+  });
```
Comment (interstella-5555) [inline src/utils.js]: I think this function doesn't make sense, it's doing exactly the same as `urljoin`.

`makeUrl('/a/', '/b/', '/c/')` === `urljoin('/a/', '/b/', '/c/')` ?

File: src/utils.js [hunk 8]
```diff
@@ -97,3 +103,16 @@ export function parseSkylink(skylink = "") {
 
   throw new Error(`Could not extract skylink from '${skylink}'`);
 }
+
+export function trimUriPrefix(str, prefix) {
+  const longPrefix = `${prefix}//`;
+  if (str.indexOf(longPrefix) == 0) {
```
Comment (interstella-5555) [inline src/utils.js]:
```suggestion
  if (str.startsWith(longPrefix)) {
```

File: src/utils.js [hunk 9]
```diff
@@ -97,3 +103,16 @@ export function parseSkylink(skylink = "") {
...
+    // longPrefix is exactly at the beginning
+    return str.slice(longPrefix.length);
+  }
+  if (str.indexOf(prefix) == 0) {
```
Comment (interstella-5555) [inline src/utils.js]:
```suggestion
  if (str.startsWith(prefix)) {
```

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/utils.js (hunk 6 above)
Comment (mrcnski) [reply src/utils.js:9]: David gave me these prefixes.

File: src/upload.js (hunk 5 above)
Comment (mrcnski) [reply src/upload.js]: I don't think so, because `makeUrl` takes a variable number of arguments. Would be nice if it was possible, though.

File: src/utils.js (hunk 7 above)
Comment (mrcnski) [reply src/utils.js]: `urljoin` failed the unit tests. I think if one of the arguments in the middle is `/` it will append it to a slash after it. I'll document this

File: src/download.js (hunk 1 above)
Comment (mrcnski) [reply src/download.js]: Hmm, it doesn't just resolve, it downloads too. Most accurate would be `resolveAndDownloadHns` but that's pretty long.

File: README.md [hunk 10]
```diff
@@ -70,10 +70,9 @@ With authentication:
 ```javascript
 import { SkynetClient } from "skynet-js";
 
-const client = new SkynetClient("https://siasky.net");
-
 async function authenticationExample() {
   try {
+    const client = new SkynetClient("https://siasky.net");
```
Comment (mrcnski) [inline README.md]: I'll revert this, wasn't intended.

File: package.json [hunk 11]
```diff
@@ -9,7 +9,7 @@
     "src/*"
   ],
   "scripts": {
-    "test": "jest",
+    "test": "jest --runInBand",
```
Comment (interstella-5555) [inline package.json:12]: any reason to run in band?

File: src/client.js [hunk 12]
```diff
@@ -1,10 +1,57 @@
...
+   * The Skynet Client which can be used to access Skynet.
+   * @constructor
+   * @param {string} [portalUrl="https://siasky.net"] - The portal URL to use to access Skynet, if specified.
+   */
+  constructor(portalUrl = "") {
+    if (portalUrl === "") {
       portalUrl = defaultPortalUrl();
     }
     this.portalUrl = portalUrl;
   }
```
Comment (interstella-5555) [inline src/client.js]:
```suggestion
  constructor(portalUrl = defaultPortalUrl()) {
    this.portalUrl = portalUrl;
  }
```

File: src/download.js [hunk 13]
```diff
@@ -1,35 +1,83 @@
...
 SkynetClient.prototype.download = function (skylink, customOptions = {}) {
-  const opts = { ...defaultDownloadOptions, ...customOptions, download: true };
+  const opts = { ...defaultDownloadOptions, ...this.customOptions, ...customOptions, download: true };
   const url = this.getDownloadUrl(skylink, opts);
 
   window.open(url, "_blank");
 };
 
+SkynetClient.prototype.downloadHns = async function (domain, customOptions = {}) {
+  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
  
  throw new Error("Unimplemented");
```

File: src/download.js [hunk 14]
```diff
@@ -1,35 +1,83 @@
...
+};
+
+SkynetClient.prototype.downloadHnsres = async function (domain, customOptions = {}) {
+  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
  
  throw new Error("Unimplemented");
```

File: src/download.js [hunk 15]
```diff
@@ -1,35 +1,83 @@
...
-SkynetClient.prototype.metadata = function (skylink, customOptions = {}) {
-  const opts = { ...defaultDownloadOptions, ...customOptions };
+SkynetClient.prototype.metadata = async function (skylink, customOptions = {}) {
+  const opts = { ...defaultDownloadOptions, ...this.customOptions, ...customOptions };
 
   throw new Error("Unimplemented");
 };
 
+/**
+ * Opens the content at the skylink within the browser.
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
 * Opens the content of the skylink within the browser.
```

File: src/download.js [hunk 16]
```diff
@@ -1,35 +1,83 @@
...
+  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
+
+  domain = trimUriPrefix(domain, uriHandshakePrefix);
+
+  // Get the skylink from the hns domain on the portal.
+  const response = await this.executeRequest({ ...opts, method: "get", extraPath: domain });
+  const skylink = response.data.skylink;
+
+  this.open(skylink, customOptions);
+};
```
Comment (interstella-5555) [inline src/download.js]: I would probably consider extending `open` to accept handshake domains with some param like `{ handshake: true }` instead of creating clones of the `open` and `download` functions

File: src/download.js [hunk 17]
```diff
@@ -1,35 +1,83 @@
...
+
+SkynetClient.prototype.openHnsres = async function (domain, customOptions = {}) {
+  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
```
Comment (interstella-5555) [inline src/download.js]:
```suggestion
  const opts = { ...defaultDownloadHnsOptions, ...this.customOptions, ...customOptions };
  
  throw new Error("Unimplemented");
```

File: src/download.test.js [hunk 18]
```diff
@@ -1,6 +1,12 @@
+import axios from "axios";
+import MockAdapter from "axios-mock-adapter";
+
 import { SkynetClient, defaultSkynetPortalUrl } from "./index";
 
+const mock = new MockAdapter(axios);
+
 const portalUrl = defaultSkynetPortalUrl;
+const hnsLink = "doesn";
```
Comment (interstella-5555) [inline src/download.test.js]: let's use foo for more general test usage

```suggestion
const hnsLink = "foo";
```

File: src/download.test.js [hunk 19]
```diff
@@ -55,3 +62,29 @@ describe("open", () => {
...
+
+      expect(mock.history.get.length).toBe(i);
+
+      expect(windowOpen).toHaveBeenCalledTimes(i);
+      expect(windowOpen).toHaveBeenCalledWith(`${portalUrl}/${skylink}`, "_blank");
+
+      i++;
+    }
+
+    mock.resetHistory();
```
Comment (interstella-5555) [inline src/download.test.js]: It's better to always reset mock before the assertions because in case of the failure, the reset might not be executed and subsequent tests will report fake success/failures. It might not be a case here but it's a good practice. Also it's a common practice in js to use function for looping array instead of an actual loop for increased safety of leaking variables out - again, not a case here but a good practice.

```suggestion
    validHnsLinkVariations.forEach((input, index) => {
      mock.resetHistory();
      
      await client.openHns(input);

      expect(mock.history.get.length).toBe(1);

      expect(windowOpen).toHaveBeenCalledTimes(1);
      expect(windowOpen).toHaveBeenCalledWith(`${portalUrl}/${skylink}`, "_blank");
    }
```

File: src/utils.js (hunk 7 above)
Comment (interstella-5555) [reply src/utils.js]: Then instead of arguments, we should use rest parameters. Usage of magic `arguments` is considered a bad practice due to it's magical nature :)

```suggestion
export function makeUrl(...args) {
  return args.reduce(function (acc, cur) {
    return urljoin(acc, cur);
  });
```

File: src/utils.js (hunk 6 above)
Comment (interstella-5555) [reply src/utils.js:9]: We don't have those documented anywhere and `uriHandshakeResolverPrefix` is not used anyway. Can we get some clarification on that @DavidVorick ?

File: src/upload.test.js [hunk 20]
```diff
@@ -1,97 +1,155 @@
...
-  it("should return single skylink on success", async () => {
-    const data = await client.uploadDirectory(directory, filename);
+  it("should send register onUploadProgress callback if defined", async () => {
+    const data = await client.uploadDirectory(directory, filename, { onUploadProgress: jest.fn() });
+
+    expect(mock.history.post.length).toBe(1);
+    const request = mock.history.post[0];
+    mock.resetHistory();
+
+    expect(request.onUploadProgress).toEqual(expect.any(Function));
```
Comment (interstella-5555) [inline src/upload.test.js:147]: you should `expect` the function that you assigned to `onUploadProgress` on the first line, not just any function

File: src/upload.test.js [hunk 21]
```diff
@@ -1,97 +1,155 @@
...
 
-    expect(axios.post).toHaveBeenCalledWith(
-      `${portalUrl}/skynet/skyfile?filename=${filename}`,
-      expect.any(FormData), // TODO: Inspect data contents.
-      undefined
-    );
-  });
+    expect(mock.history.post.length).toBe(1);
+    const request = mock.history.post[0];
+    mock.resetHistory();
```
Comment (interstella-5555) [inline src/upload.test.js]: reset history before the tests, not during (covered "why" in a different comment), same applies to other tests

File: src/upload.test.js [hunk 22]
```diff
@@ -1,97 +1,155 @@
...
+
+    expect(request.auth).toEqual({ username: "", password: "foobar" });
+    await compareFormData(request.data, [["file", "foo", filename]]);
+
+    expect(data).toEqual({ skylink });
+  });
+
+  it("should send custom user agent if defined", async () => {
+    const client2 = new SkynetClient(portalUrl, { customUserAgent: "Sia-Agent" });
```
Comment (interstella-5555) [inline src/upload.test.js]: any reason to call it `client2` and not `client`?

File: src/upload.test.js [hunk 23]
```diff
@@ -1,97 +1,155 @@
...
+
+    // Should use client's user agent.
+
+    let data = await client2.upload(file);
```
Comment (interstella-5555) [inline src/upload.test.js]:
```suggestion
    const data = await client2.upload(file);
```

File: src/upload.test.js [hunk 24]
```diff
@@ -1,97 +1,155 @@
...
+
+    expect(mock.history.post.length).toBe(1);
+    let request = mock.history.post[0];
```
Comment (interstella-5555) [inline src/upload.test.js]:
```suggestion
    const request = mock.history.post[0];
```

File: src/upload.test.js [hunk 25]
```diff
@@ -1,97 +1,155 @@
...
+    mock.resetHistory();
+
+    expect(request.headers["User-Agent"]).toEqual("Sia-Agent");
+    // Check that other headers weren't altered.
+    expect(request.headers["Content-Type"]).toEqual("application/x-www-form-urlencoded");
+    await compareFormData(request.data, [["file", "foo", filename]]);
+
+    expect(data).toEqual({ skylink });
+
+    // Should use user agent set in options to function.
```
Comment (interstella-5555) [inline src/upload.test.js]: that's a separate test

File: src/upload.test.js [hunk 26]
```diff
@@ -1,97 +1,155 @@
...
+    // Use replyOnce to catch a single request with the new URL.
+    mock.onPost(url).replyOnce(200, { skylink: skylink });
+
+    const data = await client.upload(file, { onUploadProgress: jest.fn() });
+
+    expect(mock.history.post.length).toBe(1);
+    const request = mock.history.post[0];
+    mock.resetHistory();
+
+    expect(request.onUploadProgress).toEqual(expect.any(Function));
```
Comment (interstella-5555) [inline src/upload.test.js:50]: expect the function from first line

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: package.json (hunk 11 above)
Comment (mrcnski) [reply package.json:12]: Yeah, the tests were interfering with each other otherwise. For example I use `const request = mock.history.post[0];` to get the latest request made and it only works if one test runs at a time. There might be a way around this now that I think about it, though. I'll work on this.

File: src/upload.test.js (hunk 26 above)
Comment (mrcnski) [reply src/upload.test.js:50]: Doesn't work:

```
    expect(received).toEqual(expected) // deep equality

    Expected: [Function mockConstructor]
    Received: [Function anonymous]

      48 |     mock.resetHistory();
      49 | 
    > 50 |     expect(request.onUploadProgress).toEqual(jest.fn());
         |                                      ^
```

File: src/upload.test.js (hunk 20 above)
Comment (mrcnski) [reply src/upload.test.js:147]: See other comment

File: src/upload.test.js (hunk 26 above)
Comment (interstella-5555) [reply src/upload.test.js:50]: Because you need to store result of `jest.fn()` as a variable and compare with that variable. Otherwise, these are two different objects.

File: package.json (hunk 11 above)
Comment (mrcnski) [reply package.json:12]: So I was thinking we could pass a query parameter like `?test=testName` to identify each request but it would make the tests more complicated. In band should be fine for now since the tests are quick to run.

File: src/upload.test.js (hunk 26 above)
Comment (mrcnski) [reply src/upload.test.js:50]: I tried that as well, same error

File: src/download.test.js (hunk 19 above)
Comment (mrcnski) [reply src/download.test.js]: `forEach` doesn't work if you pass it an `async` function, see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

File: src/download.js (hunk 16 above)
Comment (mrcnski) [reply src/download.js]: That looks better to me too

File: src/upload.test.js (hunk 26 above)
Comment (interstella-5555) [reply src/upload.test.js:50]: It means the test detected a real issue then.

File: package.json (hunk 11 above)
Comment (interstella-5555) [reply package.json:12]: That sounds like you leak the mocks somewhere. It shouldn't be an issue and can result in false-positives or false-negatives.

Comment (mrcnski) [reply package.json:12]: No, the problem is that the history gets scrambled with multiple requests running simultaneously, and then the test doesn't know which one to examine. Having one test run at a time and then erasing the history works fine. Open to suggestions though.

File: src/upload.test.js (hunk 26 above)
Comment (mrcnski) [reply src/upload.test.js:50]: Oh, it's because we modify `onUploadProgress` in `executeRequest`. Honestly it's enough to check here that `request.onUploadProgress` is defined and equal to any function.

**Comment by mrcnski** [comment]
Just added some more changes based on my testing.

There is still an issue with redirects causing downloads to open in the same tab. I'll still need to test downloads once that is fixed. If that works I will need to also change the download tests.

Edit: have tested downloads with Ivo's [redirect fix MR](https://gitlab.com/NebulousLabs/Sia/-/merge_requests/4773) and they work.

File: src/index.test.js [hunk 27]
```diff
@@ -5,12 +5,27 @@ describe("SkynetClient", () => {
...
     // Download
-    expect(client).toHaveProperty("download");
-    expect(client).toHaveProperty("getDownloadUrl");
-    expect(client).toHaveProperty("open");
+    expect(client).toHaveProperty("downloadFile");
+    expect(client).toHaveProperty("downloadFileHns");
+    expect(client).toHaveProperty("getHnsUrl");
+    expect(client).toHaveProperty("getHnsresUrl");
+    expect(client).toHaveProperty("getSkylinkUrl");
+    expect(client).toHaveProperty("metadata");
```
Comment (interstella-5555) [inline src/index.test.js]: this should be `getMetadata`

File: src/index.test.js [hunk 28]
```diff
@@ -5,12 +5,27 @@ describe("SkynetClient", () => {
...
+    expect(client).toHaveProperty("getSkykeyById");
+    expect(client).toHaveProperty("getSkykeyByName");
+    expect(client).toHaveProperty("getSkykeys");
 
     // Upload
-    expect(client).toHaveProperty("upload");
+    expect(client).toHaveProperty("uploadFile");
+    expect(client).toHaveProperty("uploadFileRequest");
     expect(client).toHaveProperty("uploadDirectory");
+    expect(client).toHaveProperty("uploadDirectoryRequest");
```
Comment (interstella-5555) [inline src/index.test.js:29]: why do you need to separate the request methods?

File: webpack.config.js [hunk 29]
```diff
@@ -0,0 +1,9 @@
+module.exports = {
+  mode: "development",
+  output: {
+    library: "skynet-js",
+    libraryTarget: "umd",
+    umdNamedDefine: true,
+  },
+  target: "web",
+};
```
Comment (interstella-5555) [inline webpack.config.js]: why is this needed?

File: src/index.test.js (hunk 28 above)
Comment (mrcnski) [reply src/index.test.js:29]: It's a usability impediment to have to append `.skylink` to every upload call. Since that is the desired result 99% of the time we have this as the main API as well a version of the method that returns the full request (with merkleroot and bitfield).

**Review by interstella-5555 — APPROVED** [review]
